### PR TITLE
feat: BitVec.signExtend.{toNat, toInt}

### PR DESCRIPTION
**Pull request by mhk119**
This PR adds toNat and toInt theorems for BitVec.signExtend.

### CODE REVIEW

File: src/Init/Data/BitVec/Lemmas.lean [hunk 1]
```diff
@@ -1532,6 +1536,69 @@ theorem signExtend_eq_setWidth_of_lt (x : BitVec w) {v : Nat} (hv : v ≤ w):
...
+If the msb is false, then the result equals the original value.
+If the msb is true, then we add a value of `(2^v - 2^w)`, which arises from the sign extension. -/
+theorem toNat_signExtend_of_le (x : BitVec w) {v : Nat} (hv : w ≤ v) :
+    (x.signExtend v).toNat = x.toNat + if x.msb then 2^v - 2^w else 0 := by
+  apply Nat.eq_of_testBit_eq
+  intro i
+  have ⟨k, hk⟩ := Nat.exists_eq_add_of_le hv; rw [hk]
+  rw [testBit_toNat, getLsbD_signExtend, Nat.pow_add, ← Nat.mul_sub_one, Nat.add_comm (x.toNat)]
+  by_cases hx : x.msb
+  · simp [hx, Nat.testBit_mul_pow_two_add _ x.isLt, testBit_toNat]
```
Comment (hargoniX) [inline src/Init/Data/BitVec/Lemmas.lean:1549]: Not in simp normal form.

File: src/Init/Data/BitVec/Lemmas.lean [hunk 2]
```diff
@@ -1532,6 +1536,69 @@ theorem signExtend_eq_setWidth_of_lt (x : BitVec w) {v : Nat} (hv : v ≤ w):
...
+    -- Case analysis on i being in the intervals [0..w), [w..w + k), [w+k..∞)
+    have hi : i < w ∨ (w ≤ i ∧ i < w + k) ∨ w + k ≤ i := by omega
+    rcases hi with hi | hi | hi
+    · simp [hi]; omega
+    · simp [hi]; omega
+    · simp [hi, show ¬ (i < w + k) by omega, show ¬ (i < w) by omega]
+      omega
+  · simp [hx, Nat.testBit_mul_pow_two_add _ x.isLt, testBit_toNat]
```
Comment (hargoniX) [inline src/Init/Data/BitVec/Lemmas.lean:1557]: Not in simp normal form.

File: src/Init/Data/BitVec/Lemmas.lean [hunk 3]
```diff
@@ -1532,6 +1536,69 @@ theorem signExtend_eq_setWidth_of_lt (x : BitVec w) {v : Nat} (hv : v ≤ w):
...
   rw [signExtend_eq_setWidth_of_lt _ (Nat.le_refl _), setWidth_eq]
 
+/-- Sign extending to a larger bitwidth depends on the msb.
+If the msb is false, then the result equals the original value.
+If the msb is true, then we add a value of `(2^v - 2^w)`, which arises from the sign extension. -/
+theorem toNat_signExtend_of_le (x : BitVec w) {v : Nat} (hv : w ≤ v) :
+    (x.signExtend v).toNat = x.toNat + if x.msb then 2^v - 2^w else 0 := by
+  apply Nat.eq_of_testBit_eq
+  intro i
+  have ⟨k, hk⟩ := Nat.exists_eq_add_of_le hv; rw [hk]
```
Comment (hargoniX) [inline src/Init/Data/BitVec/Lemmas.lean:1546]: A `have` followed by `; tac` where the proof continues further down feels a bit weird, maybe put the `rw` on the next line.

**Review by hargoniX — CHANGES_REQUESTED** [review]
Some minor remarks, I got two additional ones:
1. Could you rebase your branch on leanprover master? The test task IO flipper should be gone there.
2. Did you check that your theorem statements are in simp normal form?

File: src/Init/Data/BitVec/Lemmas.lean (hunk 1 above)
Comment (mhk119) [reply src/Init/Data/BitVec/Lemmas.lean:1589]: Do you mean non-terminal simps?

Comment (hargoniX) [reply src/Init/Data/BitVec/Lemmas.lean:1589]: Yes.

**Comment by hargoniX** [comment]
The PR now just seems to consist of a change to code that is fine according to the style guide? What is the motivation here?

**Comment by mhk119** [comment]
> The PR now just seems to consist of a change to code that is fine according to the style guide? What is the motivation here?

My bad, I didn't see that the theorems got merged into main in the meantime. Closing this :).

**Comment by tobiasgrosser** [comment]
It seems these got added in https://github.com/leanprover/lean4/pull/6155 and https://github.com/leanprover/lean4/pull/6157. In the future, we should cross-reference commits when someone decides to split them. Thank you @mhk119 for getting this in.

**Pull request closed without merging**